src/lib/system/datasources/local_music_fetcher_impl.dart

```diff
@@ -30,200 +33,306 @@ class LocalMusicFetcherImpl implements LocalMusicFetcher {
-              } else {
-                colorMap[album.id] = album.color;
-              }
-            }
-            albums.add(album.copyWith(color: color));
-            final artist = artistsInDb.singleWhere((a) => a.name == album.artist);
-            artistSet.add(artist);
-          } else {
-            // we already encountered the album (at least by albumString)
-            // make sure the id is consistent
-            if (album.id != albumIdMap[albumString]) {
-              songs.add(
-                song.copyWith(
-                  albumId: albumIdMap[albumString],
-                  color: colorMap[albumIdMap[albumString]],
-                ),
-              );
-              continue;
-            }
-          }
-          songs.add(song.copyWith(color: colorMap[album.id]));
-          continue;
-        } else {
-          // read new info but keep albumId
-          albumId = song.albumId;
-        }
-      }
+    final albumIdMap = structs['albumIdMap'] as Map<String, int>;
+    final albumArtMap = structs['albumArtMap'] as Map<int, String>;
 
-      final Metadata songData;
-      try {
-        songData = await MetadataGod.readMetadata(file: songFile.path);
-      } on FfiException {
-        continue;
-      }
+    final songsToCheck = await getMetadataForFiles(songFilesToCheck);
 
-      // completely new song -> new album ids should start after existing ones
-      // this is new information
-      // is the album ID still correct or do we find another album with the same properties?
-      final String albumArtist = songData.albumArtist ?? '';
-      albumString = '${songData.album}___${albumArtist}__${songData.year}';
-
-      String? albumArtPath;
-      if (!albumIdMap.containsKey(albumString)) {
-        // we haven't seen an album with these properties in the files yet, but there might be an entry in the database
-        // in this case, we should use the corresponding ID
-        albumId ??= await _musicDataSource.getAlbumId(
-              songData.album,
-              albumArtist,
-              songData.year,
-            ) ??
-            newAlbumId++;
-        albumIdMap[albumString] = albumId;
-
-        final albumArt = songData.picture;
-
-        if (albumArt != null) {
-          albumArtPath = '${dir.path}/$albumId';
-          final file = File(albumArtPath);
-          file.writeAsBytesSync(albumArt.data);
-          albumArtMap[albumId] = albumArtPath;
-
-          color = await getBackgroundColor(FileImage(file));
-          colorMap[albumId] = color;
-        }
+    for (final (songFile, songData) in songsToCheck) {
+      _log.i('Scanning Song ${songFile.path}');
+
+      final lastModified = songFile.lastModifiedSync();
 
-        final String songArtist = songData.artist ?? '';
-        final String artistName =
-            albumArtist != '' ? albumArtist : (songArtist != '' ? songArtist : DEF_ARTIST);
+      final albumArtist = songData.albumArtist;
```

Comment:
Is this guaranteed to yield an artist? Sometimes albumArtist is not available and we have to use the "regular" artist instead.